# Практическое задание №4 по теме "Непрерывные случайные величины. Функция распределения и функция плотности. Нормальное распределение. Центральная предельная теорема".

In [1]:
import numpy as np
from scipy import stats
from scipy import integrate

from matplotlib import pyplot as plt
plt.style.use('seaborn-whitegrid')
%config InlineBackend.figure_formats = ['svg']

## Задача 1

Случайная непрерывная величина A имеет равномерное распределение на промежутке (200, 800].
Найдите ее среднее значение и дисперсию.

___Решение___

1-ый способ:

In [2]:
a = stats.uniform.rvs(loc=200, scale=600, size=100000000)

In [3]:
a_mean = a.mean()
a_var = a.var()
print(f"Среднее значение: {np.round(a_mean, 5)}\nДисперсия: {np.round(a_var, 5)}")

Среднее значение: 500.00317
Дисперсия: 29999.27


2-ой способ:

$$M(X) = \displaystyle\int_{-\infty}^\infty x \cdot f(x) \: dx$$ 

$$D(X) = \displaystyle\int_{-\infty}^\infty (x - M(X))^2 \cdot f(x) \: dx$$

$$f(x) = \frac{1}{b-a}$$

Численно

In [4]:
def f_mx(x):
    return 1/(b-a)*x

a=200
b=800
MX, err = integrate.quad(f_mx, a, b)

In [5]:
def f_dx(x):
    return (x-MX)**2/(b-a)
DX, err = integrate.quad(f_dx, a, b)

In [6]:
print (f"Мат.ожидание M={round(MX)}, Дисперсия D={round(DX)}")

Мат.ожидание M=500, Дисперсия D=30000


Аналитически

In [7]:
from sympy import *

In [8]:
a=200
b=800
x=Symbol('x')
f_mx = 1/(b-a)*x
MX = integrate(f_mx,(x, a, b))
print (f"Мат.ожидание M={round(MX)}")

Мат.ожидание M=500


In [9]:
f_dx = (x-MX)**2/(b-a)
DX = integrate(f_dx,(x, a, b))
print (f"Мат.ожидание D={round(DX)}")

Мат.ожидание D=30000


3-ий способ:

$$f(x)=1/(b-a)$$

$$M(X) = \displaystyle\int_{a}^{b} x \cdot f(x) \: dx = \displaystyle\frac{x^2}{2(b-a)}\bigg|_{a}^{b} = \frac{b+a}{2}$$

$$D(X) = \displaystyle\int_{a}^{b} (x - M(X))^2 \cdot f(x) \: dx = \displaystyle\frac{(x-M(X))^3}{3(b-a)}\bigg|_{a}^{b} = \frac{(b-\frac{a+b}{2})^3}{3(b-a)} - \frac{(a-\frac{a+b}{2})^3}{3(b-a)} = $$

$$ = \frac{(b-a)^3}{24(b-a)} + \frac{(a-b)^3}{24(a-b)} = \frac{(b-a)^2}{24} + \frac{(b-a)^2}{24} = \frac{(b-a)^2}{12}$$

In [10]:
b = 800
a = 200
MX = (b+a)/2
DX = (b-a)**2/12
print (f"Мат.ожидание M={round(MX)}, Дисперсия D={round(DX)}")

Мат.ожидание M=500, Дисперсия D=30000


___Ответ:___ Среднее значение: 500, Дисперсия: 30000

---

## Задача 2

О случайной непрерывной равномерно распределенной величине B известно, что ее дисперсия равна 0.2. Можно ли найти правую границу величины B и ее среднее значение зная, что левая граница равна 0.5? Если да, найдите ее.

___Решение___

1-ый способ:

In [11]:
right_border = 0.5
left_border = 0.5
k = 3
step = np.power(0.1, k)
n = 1000000
result = 0
for i in range(n):
    right_border += step
    b = np.random.uniform(left_border, right_border, 1000000)
    b_var = b.var(ddof=1)
    if (b_var >= 0.2):
        result = right_border
        break
print(f"Правая граница величины B: {np.round(result, k)}\nДисперсия: {np.round(b_var, k)}")

Правая граница величины B: 2.048
Дисперсия: 0.2


Проверка:

In [12]:
b = np.random.uniform(0.5, 2.05, 10000000)
b.var(ddof=1)

0.20023180841244156

In [13]:
# Среднее значение
np.round(b.mean(), 3)

1.275

2-ой способ:

$$a = 0.5$$

$$D(X) = \frac{(b-a)^2}{12} = 0.2$$

Можно составить уравнение: 
$$\frac{(b-0.5)^2}{12} = 0.2$$

$$(b-0.5-\sqrt{2.4})(b-0.5+\sqrt{2.4})=0$$

$$\begin{gathered}
b = 0.5 + \sqrt{2.4} \\
b = 0.5 - \sqrt{2.4}
\end{gathered}\,\,\,\Rightarrow\,\,\,
\begin{gathered}
b = 2.049193338482967 \\
b = -1.0491933384829668
\end{gathered}$$

In [14]:
DX = 0.2
a = 0.5

In [15]:
b1 = 0.5 + np.sqrt(2.4) 
b2 = 0.5 - np.sqrt(2.4)
(b1, b2)

(2.049193338482967, -1.0491933384829668)

In [16]:
b = b1
MX = (b+a)/2

In [17]:
print(f"Правая граница величины B: {np.round(b, 5)}\nСреднее значение: {np.round(MX, 5)}")

Правая граница величины B: 2.04919
Среднее значение: 1.2746


___Ответ:___ Правая граница: 2.04919, Среднее значение: 1.2746

---

## Задача 3

Непрерывная случайная величина X распределена нормально и задана плотностью распределения $f(x)=\frac{1}{4\sqrt{2\pi}}\cdot\frac{e^{-(x+2)^{2}}}{32}$

Найдите:

    а) M(X)
    б) D(X)
    в) std(X) (среднее квадратичное отклонение)

___Решение___

Плотность вероятности нормального распределения:
$$f(x) = \dfrac{1}{\sqrt{2 \pi \sigma^2}} \cdot \exp \left(- \frac{(x - \mu)^2}{2 \sigma^2} \right)$$

Выражение $f(x)=\frac{1}{4\sqrt{2\pi}}\cdot\frac{e^{-(x+2)^{2}}}{32}$ соответствует нормальному распределнию с параметрами $\mu = -2$ и $\sigma = 4$ 

In [18]:
mu = -2
sigma = 4
X = stats.norm(loc=mu, scale=sigma)

In [19]:
X_mean = X.mean()
X_var = X.var()
X_std = X.std()

In [20]:
X_mean, X_var, X_std

(-2.0, 16.0, 4.0)

Другой способ в виде интегрирования по определению MX, DX непрерывной случайной величины:

$$M(X) = \displaystyle\int_{a}^{b} x \cdot f(x) \: dx$$

$$D(X) = \displaystyle\int_{a}^{b} (x - M(X))^2 \cdot f(x) \: dx $$

In [21]:
from scipy import integrate

In [22]:
mu = -2
sigma = 4
pi = np.pi
a = -200
b = 200  

In [23]:
def f_mx(x):
    return x * (1/np.sqrt(2*pi*sigma**2)) * np.exp(-((x-mu)**2) / (2*sigma**2))
                                        
MX, err = integrate.quad(f_mx, a, b)
np.round(MX, 8)

-2.0

In [24]:
def f_dx(x):
    return  ((x-MX)**2) * (1/np.sqrt(2*pi*sigma**2)) * np.exp(-((x-mu)**2) / (2*sigma**2))
                                        
DX, err = integrate.quad(f_dx, a, b)
np.round(DX, 8)

16.0

In [25]:
std = np.sqrt(DX)
std

4.0

___Ответ:___ _а)_ $M(X) = -2$; _б)_ $D(X) = 16$; _в)_ $std(X) = 4$

---

## Задача 4

Рост взрослого населения города X имеет нормальное распределение. Причем, средний рост равен 174 см, а среднее квадратичное отклонение равно 8 см. Какова вероятность того, что случайным образом выбранный взрослый человек имеет рост:

    а) больше 182 см
    б) больше 190 см
    в) от 166 см до 190 см
    г) от 166 см до 182 см
    д) от 158 см до 190 см
    е) не выше 150 см или не ниже 190 см
    ё) не выше 150 см или не ниже 198 см
    ж) ниже 166 см.


___Решение___

Параметры нормального распределения из условия задания: $\mu = 174$ и $\sigma = 8$ 

In [26]:
mu = 174
sigma = 8
X = stats.norm(loc=mu, scale=sigma)

In [27]:
a_ = 1 - X.cdf(182)
b_ = 1 - X.cdf(190)
v_ = X.cdf(190) - X.cdf(166)
g_ = X.cdf(182) - X.cdf(166)
d_ = X.cdf(190) - X.cdf(158)
e_ = X.cdf(150) + (1 - X.cdf(190))
i_ = X.cdf(150) + (1 - X.cdf(198))
j_ = X.cdf(166)

In [28]:
print (f"a) x>182 Вероятность P={a_}")
print (f"б) x>190 Вероятность P={b_}")
print (f"в) 166<x<190 Вероятность P={v_}")
print (f"г) 166<x<182 Вероятность P={g_}")
print (f"д) 158<x<190 Вероятность P={d_}")
print (f"е) x<150 или x>190 Вероятность P={e_}")
print (f"ё) x<150 или x>198 Вероятность P={i_}")
print (f"ж) x<166 Вероятность P={j_}")

a) x>182 Вероятность P=0.15865525393145707
б) x>190 Вероятность P=0.02275013194817921
в) 166<x<190 Вероятность P=0.8185946141203637
г) 166<x<182 Вероятность P=0.6826894921370859
д) 158<x<190 Вероятность P=0.9544997361036416
е) x<150 или x>190 Вероятность P=0.0241000299798093
ё) x<150 или x>198 Вероятность P=0.0026997960632601965
ж) x<166 Вероятность P=0.15865525393145707


In [29]:
result = (a_, b_, v_, g_, d_, e_, i_, j_)
result = np.round(result, 5)

In [30]:
result

array([0.15866, 0.02275, 0.81859, 0.68269, 0.9545 , 0.0241 , 0.0027 ,
       0.15866])

Другой способ в виде интегрирования по определению:

$$N(x) = \frac{1}{8\cdot\sqrt{2\pi}}\cdot e^{-\frac{(x-174)^2}{2\cdot64}}$$

$$Ф(a,b) = \int_{a}^{b} N(x) \: dx $$ 

In [31]:
def FF(x):
    return   (1/np.sqrt(2*pi*sigma**2)) * np.exp(-((x-mu)**2) / (2*sigma**2))

mu = 174
sigma = 8

In [32]:
# a)
FF_a, err = integrate.quad(FF, 182, 1000)
np.round(FF_a, 8)

0.15865525

In [33]:
# б)
FF_b, err = integrate.quad(FF, 190, 1000)
np.round(FF_b, 8)

0.02275013

In [34]:
# в)
FF_v, err = integrate.quad(FF, 166, 190)
np.round(FF_v, 8)

0.81859461

In [35]:
# г)
FF_g, err = integrate.quad(FF, 166, 182)
np.round(FF_g, 8)

0.68268949

In [36]:
# д)
FF_d, err = integrate.quad(FF, 158, 190)
np.round(FF_d, 8)

0.95449974

In [37]:
# е)
FF_e_1, err = integrate.quad(FF, -1000, 150) 
FF_e_2, err = integrate.quad(FF, 190, 1000)
FF_e = FF_e_1 + FF_e_2
np.round(FF_e, 8)

0.02410003

In [38]:
# ё)
FF_i_1, err = integrate.quad(FF, -1000, 150) 
FF_i_2, err = integrate.quad(FF, 198, 1000)
FF_i = FF_i_1 + FF_i_2
np.round(FF_i, 8)

0.0026998

In [39]:
# ж)
FF_a, err = integrate.quad(FF, -1000, 166)
np.round(FF_a, 8)

0.15865525

___Ответ:___

    а) 0.15866,
    б) 0.02275,
    в) 0.81859,
    г) 0.68269,
    д) 0.9545,
    е) 0.0241,
    ё) 0.0027,
    ж) 0.15866

---

## Задача 5

На сколько сигм (средних квадратичных отклонений) отклоняется рост человека, равный 190 см, от математического ожидания роста в популяции, в которой $M(X)$ = 178 см и $D(X)$ = 25 кв.см?

___Решение___

$\sigma = \sqrt{D(X)}$ $$\frac{|190-M(X)|} {\sigma}= \frac{190-178} {5} = 2.4$$

In [41]:
value = 190
mx = 178
dx = 25
sigma = np.sqrt(dx)
deviation = np.abs(value - mx)/sigma 
deviation

2.4

___Ответ:___ Отклонение на $2.4\sigma$

---